File: Data Structures & Algorithms/kth-smallest-integer-in-bst/submission-2.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def kthSmallest(self, root: Optional[TreeNode], k: int) -> int:
        if not root:
            return None
        
        cnt = [k]
        ans = [-1]
        self.dfsInorder(root,cnt, ans)
        return ans[0]
    
    def dfsInorder(self, root, cnt, ans):
        if not root:
            return None
          
        self.dfsInorder(root.left, cnt, ans)
        
        cnt[0] -= 1
        if cnt[0] == 0:
           ans[0] = root.val

        self.dfsInorder(root.right, cnt, ans) 

File: Data Structures & Algorithms/binary-tree-right-side-view/submission-2.cs
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int val=0, TreeNode left=null, TreeNode right=null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

public class Solution {
    public List<int> RightSideView(TreeNode root) {
        List<int> ans = new List<int>();
       
        //BFS but start right and take whats possible
        Queue<TreeNode> queue = new Queue<TreeNode>();

        if(root != null){
            queue.Enqueue(root);
        }

        while(queue.Count > 0){
            int levelLenght = queue.Count;
            List<int> row = new List<int>();
            
            for(int i=0; i<levelLenght; ++i){
                //Process current node -> take first 1 to be
                TreeNode curr = queue.Dequeue();
                row.Add(curr.val);

                //Add Children right to left
                if(curr.right != null){
                    queue.Enqueue(curr.right);
                }
                if(curr.left != null){
                    queue.Enqueue(curr.left);
                }
            }

            //Add only the 1st element of the row
            ans.Add(row[0]);
        }

        return ans;
    }
}
